File: svc-data/src/test/java/com/beancounter/marketdata/integ/MarketMvcTests.java
package com.beancounter.marketdata.integ;

import static com.beancounter.marketdata.utils.RegistrationUtils.registerUser;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.jwt;
import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.beancounter.auth.common.TokenUtils;
import com.beancounter.auth.server.AuthorityRoleConverter;
import com.beancounter.common.contracts.MarketResponse;
import com.beancounter.common.exception.BusinessException;
import com.beancounter.common.model.Market;
import com.beancounter.common.model.SystemUser;
import com.beancounter.common.utils.BcJson;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

@ExtendWith(SpringExtension.class)
@SpringBootTest
@ActiveProfiles("test")
@Tag("slow")
class MarketMvcTests {

  private final ObjectMapper objectMapper = new BcJson().getObjectMapper();
  private final AuthorityRoleConverter authorityRoleConverter = new AuthorityRoleConverter();
  @Autowired
  private WebApplicationContext wac;
  private MockMvc mockMvc;
  private Jwt token;

  @BeforeEach
  void mockServices() {
    this.mockMvc = MockMvcBuilders.webAppContextSetup(wac)
        .apply(springSecurity())
        .build();
    SystemUser user = new SystemUser("MarketMvcTests", "MarketMvcTests@testing.com");

    token = TokenUtils.getUserToken(user);
    registerUser(mockMvc, token);

  }

  @Test
  void is_AllMarketsFound() throws Exception {

    MvcResult mvcResult = mockMvc.perform(
        get("/markets/")
            .with(jwt().jwt(token).authorities(authorityRoleConverter))
            .contentType(MediaType.APPLICATION_JSON)
    ).andExpect(status().isOk())
        .andExpect(content().contentType(MediaType.APPLICATION_JSON))
        .andReturn();

    MarketResponse marketResponse = objectMapper
        .readValue(mvcResult.getResponse().getContentAsString(), MarketResponse.class);
    assertThat(marketResponse).isNotNull().hasFieldOrProperty("data");
    assertThat(marketResponse.getData()).isNotEmpty();
  }

  @Test
  void is_SingleMarketFoundCaseInsensitive() throws Exception {
    MvcResult mvcResult = mockMvc.perform(
        get("/markets/nzx")
            .with(jwt().jwt(token).authorities(authorityRoleConverter))
            .contentType(MediaType.APPLICATION_JSON)
    ).andExpect(status().isOk())
        .andExpect(content().contentType(MediaType.APPLICATION_JSON))
        .andReturn();

    MarketResponse marketResponse = objectMapper
        .readValue(mvcResult.getResponse().getContentAsString(), MarketResponse.class);
    assertThat(marketResponse).isNotNull().hasFieldOrProperty("data");
    assertThat(marketResponse.getData()).isNotNull().hasSize(1);
    Market nzx = marketResponse.getData().iterator().next();
    assertThat(nzx).hasNoNullFieldsOrPropertiesExcept("currencyId", "timezoneId", "enricher");
  }

  @Test
  void is_SingleMarketBadRequest() throws Exception {
    ResultActions result = mockMvc.perform(
        get("/markets/non-existent")
            .with(jwt().jwt(token).authorities(authorityRoleConverter))
            .contentType(MediaType.APPLICATION_JSON)
    ).andExpect(status().is4xxClientError());

    assertThat(result.andReturn().getResolvedException())
        .isNotNull()
        .isInstanceOfAny(BusinessException.class);

  }

}

File: svc-data/src/test/java/com/beancounter/marketdata/integ/MarketMvcTests.kt
package com.beancounter.marketdata.integ

import com.beancounter.auth.common.TokenUtils
import com.beancounter.auth.server.AuthorityRoleConverter
import com.beancounter.common.contracts.MarketResponse
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.model.SystemUser
import com.beancounter.common.utils.BcJson
import com.beancounter.marketdata.utils.RegistrationUtils.registerUser
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.DefaultMockMvcBuilder
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext

@ExtendWith(SpringExtension::class)
@SpringBootTest
@ActiveProfiles("test")
@Tag("slow")
internal class MarketMvcTests {
    private val objectMapper = BcJson().objectMapper
    private val authorityRoleConverter = AuthorityRoleConverter()

    @Autowired
    private lateinit var wac: WebApplicationContext
    private lateinit var mockMvc: MockMvc
    private var token: Jwt = TokenUtils()
        .getUserToken(SystemUser("MarketMvcTests", "MarketMvcTests@testing.com"))

    @BeforeEach
    fun mockServices() {
        mockMvc = MockMvcBuilders.webAppContextSetup(wac)
            .apply<DefaultMockMvcBuilder>(SecurityMockMvcConfigurers.springSecurity())
            .build()
        registerUser(mockMvc, token)
    }

    @Test
    @Throws(Exception::class)
    fun is_AllMarketsFound() {
        val mvcResult = mockMvc.perform(
            MockMvcRequestBuilders.get("/markets/")
                .with(SecurityMockMvcRequestPostProcessors.jwt().jwt(token).authorities(authorityRoleConverter))
                .contentType(MediaType.APPLICATION_JSON)
        ).andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andReturn()
        val marketResponse = objectMapper
            .readValue(mvcResult.response.contentAsString, MarketResponse::class.java)
        Assertions.assertThat(marketResponse).isNotNull.hasFieldOrProperty("data")
        Assertions.assertThat(marketResponse.data).isNotEmpty
    }

    @Test
    @Throws(Exception::class)
    fun is_SingleMarketFoundCaseInsensitive() {
        val mvcResult = mockMvc.perform(
            MockMvcRequestBuilders.get("/markets/nzx")
                .with(SecurityMockMvcRequestPostProcessors.jwt().jwt(token).authorities(authorityRoleConverter))
                .contentType(MediaType.APPLICATION_JSON)
        ).andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
            .andReturn()
        val marketResponse = objectMapper
            .readValue(mvcResult.response.contentAsString, MarketResponse::class.java)
        Assertions.assertThat(marketResponse).isNotNull.hasFieldOrProperty("data")
        Assertions.assertThat(marketResponse.data).isNotNull.hasSize(1)
        val nzx = marketResponse.data!!.iterator().next()
        Assertions.assertThat(nzx).hasNoNullFieldsOrPropertiesExcept("currencyId", "timezoneId", "enricher")
    }

    @Test
    @Throws(Exception::class)
    fun is_SingleMarketBadRequest() {
        val result = mockMvc.perform(
            MockMvcRequestBuilders.get("/markets/non-existent")
                .with(SecurityMockMvcRequestPostProcessors.jwt().jwt(token).authorities(authorityRoleConverter))
                .contentType(MediaType.APPLICATION_JSON)
        ).andExpect(MockMvcResultMatchers.status().is4xxClientError)
        Assertions.assertThat(result.andReturn().resolvedException)
            .isNotNull
            .isInstanceOfAny(BusinessException::class.java)
    }
}
